### PR TITLE
Quick Deploy v0

### PR DESCRIPTION
**Pull request by jamcohen**
One way to reduce the size of a Unity-based instant app is to place some assets in an AssetBundle and download that AssetBundle from a loading screen at game startup.

Quick Deploy makes it easier to use AssetBundles in this fashion via two main features:
- It simplifies the creation of an AssetBundle from existing scenes.
- It provides a loading screen that displays progress in downloading the AssetBundle from a server.

### CODE REVIEW

File: GooglePlayInstant/Editor/PlayInstantEditorMenu.cs [hunk 1]
```diff
@@ -68,21 +69,31 @@ private static void SetUpPlayInstantSdk()
             PlayInstantSdkInstaller.SetUp();
         }
 
-#if UNITY_2018_3_OR_NEWER
-        [MenuItem("PlayInstant/Build Android App Bundle...", false, 300)]
-        private static void BuildAndroidAppBundle()
+        [MenuItem("PlayInstant/Quick Deploy/Overview...", false, 300)]
```
Comment (thanwah) [inline GooglePlayInstant/Editor/PlayInstantEditorMenu.cs]: Should Quick Deploy be moved into the previous section, e.g. as "203"? It kind of seem strange to me to have the horizontal dividers above and below a single menu item.

File: GooglePlayInstant/LoadingScreen/LoadingBar.cs [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+                SetProgress(Progress);
+            }
+        }
+
+        // TODO: Make sure this scales correctly in landscape
+        public void ApplyBorderWidth()
+        {
+            if (_rectTransform == null)
+            {
+                _rectTransform = GetComponent<RectTransform>();
```
Comment (thanwah) [inline GooglePlayInstant/LoadingScreen/LoadingBar.cs]: Given that _rectTransform isn't used in the remainder of this method, is this if block needed?

File: GooglePlayInstant/LoadingScreen/LoadingScreen.cs [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+        private IEnumerator HandleRequestFailed(string assetBundleUrl, UnityWebRequest webRequest)
+        {
+            if (_assetBundleRetrievalAttemptCount < MaxAttemptCount)
+            {
+                _assetBundleRetrievalAttemptCount++;
+                Debug.LogFormat("Attempt #{0} at downloading AssetBundle...", _assetBundleRetrievalAttemptCount);
+                yield return new WaitForSeconds(2);
+
+                //TODO: revisit this methodology of setting the loading bar
+                LoadingBar.SetProgress(0f);
```
Comment (thanwah) [inline GooglePlayInstant/LoadingScreen/LoadingScreen.cs]: I thought we decided that progress should never go backwards?

File: GooglePlayInstant/Editor/PlayInstantEditorMenu.cs [hunk 4]
```diff
@@ -68,21 +69,31 @@ private static void SetUpPlayInstantSdk()
             PlayInstantSdkInstaller.SetUp();
         }
 
-#if UNITY_2018_3_OR_NEWER
```
[the remark this answers is not in the document]
Comment (jamcohen) [reply GooglePlayInstant/Editor/PlayInstantEditorMenu.cs]: Resolved

File: GooglePlayInstant/Editor/QuickDeploy/DialogHelper.cs [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+using System.IO;
+using UnityEditor;
+using UnityEngine;
+
+namespace GooglePlayInstant.Editor.QuickDeploy
+{
+    /// <summary>
+    /// Utility class for displaying popup window messages.
+    /// </summary>
+    public static class DialogHelper
```
[the remark this answers is not in the document]
Comment (jamcohen) [reply GooglePlayInstant/Editor/QuickDeploy/DialogHelper.cs]: Good call, I'll add those in a separate PR

File: GooglePlayInstant/Editor/QuickDeploy/QuickDeployWindow.cs [hunk 6]
```diff
@@ -0,0 +1,353 @@
...
+                    OnGuiLoadingScreenSelect();
+                    break;
+            }
+        }
+
+        /// <summary>
+        /// Unfocus the window if the user has just moved to a different quick deploy tab.
+        /// </summary>
+        /// <param name="currentTab">A ToolBarSelectedButton instance representing the current quick deploy tab.</param>
+        /// <see cref="b/112536394"/>
```
[the remark this answers is not in the document]
Comment (jamcohen) [reply GooglePlayInstant/Editor/QuickDeploy/QuickDeployWindow.cs]: Resolved

File: GooglePlayInstant/LoadingScreen/LoadingBar.cs (hunk 2 above)
Comment (jamcohen) [reply GooglePlayInstant/LoadingScreen/LoadingBar.cs]: Good catch, that variable isn't used at all anymore.

File: GooglePlayInstant/LoadingScreen/LoadingScreen.cs (hunk 3 above)
Comment (jamcohen) [reply GooglePlayInstant/LoadingScreen/LoadingScreen.cs]: I had planned to update that when we added the retry button, but for now I'll make it resume from it's last maximum fill.

File: GooglePlayInstant/Editor/PlayInstantEditorMenu.cs (hunk 1 above)
Comment (jamcohen) [reply GooglePlayInstant/Editor/PlayInstantEditorMenu.cs]: That's fair, I'll put it in the same section as "Build Settings..." and "Set Up Instant Apps Development SDK..."

**Comment by thanwah** [comment]
Hit this error during local testing. Not sure if I can repro yet.

```
InvalidOperationException: Operation is not valid due to the current state of the object
System.Collections.Generic.Stack`1[System.Boolean].Pop ()
UnityEditor.EditorGUI.EndChangeCheck () (at /Users/builduser/buildslave/unity/build/Editor/Mono/EditorGUI.cs:315)
GooglePlayInstant.Editor.QuickDeploy.QuickDeployWindow.OnGuiLoadingScreenSelect () (at Assets/play-instant-unity-plugin/GooglePlayInstant/Editor/QuickDeploy/QuickDeployWindow.cs:288)
GooglePlayInstant.Editor.QuickDeploy.QuickDeployWindow.OnGUI () (at Assets/play-instant-unity-plugin/GooglePlayInstant/Editor/QuickDeploy/QuickDeployWindow.cs:106)
System.Reflection.MonoMethod.Invoke (System.Object obj, BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) (at /Users/builduser/buildslave/mono/build/mcs/class/corlib/System.Reflection/MonoMethod.cs:222)
Rethrow as TargetInvocationException: Exception has been thrown by the target of an invocation.
System.Reflection.MonoMethod.Invoke (System.Object obj, BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) (at /Users/builduser/buildslave/mono/build/mcs/class/corlib/System.Reflection/MonoMethod.cs:232)
System.Reflection.MethodBase.Invoke (System.Object obj, System.Object[] parameters) (at /Users/builduser/buildslave/mono/build/mcs/class/corlib/System.Reflection/MethodBase.cs:115)
UnityEditor.HostView.Invoke (System.String methodName, System.Object obj) (at /Users/builduser/buildslave/unity/build/Editor/Mono/HostView.cs:295)
UnityEditor.HostView.Invoke (System.String methodName) (at /Users/builduser/buildslave/unity/build/Editor/Mono/HostView.cs:288)
UnityEditor.HostView.OldOnGUI () (at /Users/builduser/buildslave/unity/build/Editor/Mono/HostView.cs:104)
UnityEngine.Experimental.UIElements.IMGUIContainer.DoOnGUI (UnityEngine.Event evt, Boolean isComputingLayout) (at /Users/builduser/buildslave/unity/build/Modules/UIElements/IMGUIContainer.cs:215)

```

**Pull request merged**